src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Utf8.cs

```diff
@@ -0,0 +1,83 @@
+{
+    public partial class CompareInfo
+    {
+        /// <summary>
+        /// Determines whether a UTF-8 string starts with a specific prefix.
+        /// </summary>
+        /// <param name="source">The UTF-8 string to search within.</param>
+        /// <param name="prefix">The prefix to attempt to match at the start of <paramref name="source"/>.</param>
+        /// <param name="options">The <see cref="CompareOptions"/> to use during the match.</param>
+        /// <returns>
+        /// <see langword="true"/> if <paramref name="prefix"/> occurs at the start of <paramref name="source"/>;
+        /// otherwise, <see langword="false"/>.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// <paramref name="options"/> contains an unsupported combination of flags.
+        /// </exception>
+        internal bool IsPrefixUtf8(ReadOnlySpan<byte> source, ReadOnlySpan<byte> prefix, CompareOptions options = CompareOptions.None)
+        {
+            // The empty UTF-8 string is trivially a prefix of every other string. For compat with
+            // earlier versions of the Framework we'll early-exit here before validating the
+            // 'options' argument.
+
+            if (prefix.IsEmpty)
+            {
+                return true;
+            }
+
+            if ((options & ValidIndexMaskOffFlags) == 0)
+            {
+                // Common case: caller is attempting to perform a linguistic search.
+                // Pass the flags down to NLS or ICU unless we're running in invariant
+                // mode, at which point we normalize the flags to Ordinal[IgnoreCase].
+
+                if (!GlobalizationMode.Invariant)
+                {
+                    return StartsWithCoreUtf8(source, prefix, options);
+                }
+
+                if ((options & CompareOptions.IgnoreCase) == 0)
+                {
+                    return source.StartsWith(prefix);
+                }
+
+                return source.StartsWithOrdinalIgnoreCaseUtf8(prefix);
+            }
+            else
+            {
+                // Less common case: caller is attempting to perform non-linguistic comparison,
+                // or an invalid combination of flags was supplied.
+
+                if (options == CompareOptions.Ordinal)
+                {
+                    return source.StartsWith(prefix);
+                }
+
+                if (options == CompareOptions.OrdinalIgnoreCase)
+                {
+                    return source.StartsWithOrdinalIgnoreCaseUtf8(prefix);
+                }
+
+                ThrowCompareOptionsCheckFailed(options);
+
+                return false; // make the compiler happy;
+            }
+        }
+
+        private unsafe bool StartsWithCoreUtf8(ReadOnlySpan<byte> source, ReadOnlySpan<byte> prefix, CompareOptions options)
+        {
+            // NLS/ICU doesn't provide native UTF-8 support so we need to convert to UTF-16 and compare that way
```

Comment:
I'm surprised ICU doesn't.  What do other platforms do where they're main string abstractions are implemented as UTF8?

Reply:
Right, I was as well. `libICU` has a page covering UTF-8 here and that there are only a small number of functions which actually work with UTF-8: https://unicode-org.github.io/icu/userguide/strings/utf-8.html

They recommend conversion using their own functions for performance. For the "worker functions" which do natively support UTF-8, I didn't see ones that covered our current internal wrappers: https://github.com/dotnet/runtime/blob/1af3c9eb5d97b61f75a387e43d9c9d6a20dd2e23/src/native/libs/System.Globalization.Native/pal_collation.c#L1246

It may be possible to do, but would require significantly more work to get functioning end to end and tested. I believe it may be possible if we directly use the C++ surface area and the `UTF8CollationIterator`

Reply:
Culture-aware operations (such as this one) are painfully slow in UTF-8. All of the lookup tables are written with an eye toward UTF-16, since that's where you're going to find the best performance. So while ICU does offer some rudimentary UTF-8 support here, it's not really their bread and butter.

> What do other platforms do where their main string abstractions are implemented as UTF8?

.NET is unique in that we're the only large platform that pushes culture-aware string operations so heavily by default. If you squint a bit, you can kind of see Swift doing this as well in limited cases, but they perform some UTF-16 bookmarking under the covers to keep everything solid. Other platforms - Java, PHP, Python, etc. - bury culture aware comparisons under an obscure API surface since they're so rarely used. Performance of culture-aware operations isn't a top-of-mind issue for those frameworks.